### PR TITLE
Refresh addon manifests when same-version content changes

### LINKED ISSUE #2466: [Bug]: Addon manifest cache does not refresh same-version catalog changes
### Pre-flight checks

- [x] I searched existing issues and this is not a duplicate.
- [x] I can reproduce this on the latest release or latest main build.
- [x] This issue is limited to a single bug (not multiple unrelated problems).
- [x] This is not a source/stream-specific error (the issue happens regardless of which source is used).

### App version / commit

dev@5623e12 / 0.7.11-beta

### Install method

Built from source

### Platform

Android TV

### Device model

Not device-specific; repository/cache behavior

### Android version

Not device-specific

### Player mode

Internal player

### Area (tag)

Other

### Bug description

NuvioTV can keep using stale addon catalog definitions when an installed Stremio addon changes meaningful manifest content without changing the top-level manifest `version`. This affects configurable addons where catalog visibility, catalog lists, config versions, poster URLs, paging, extras, ordering, or addon labels can change for the same manifest URL. The result is that home rows and addon reads can continue using old catalog definitions even after a stale-manifest refresh should have picked up the changed manifest.

### Steps to reproduce

1. Add a Stremio addon that exposes catalogs in its manifest.
2. Let NuvioTV load the addon and cache its manifest.
3. Update the same manifest URL so catalog definitions change while the manifest `version` stays the same. Examples include changing catalog `showInHome` flags, catalog poster URLs, catalog extras, page size, catalog order, config version, or addon label.
4. Restart NuvioTV or wait for the manifest cache to become stale so the app fetches the manifest again.
5. Open the home screen or use a one-shot addon read path that depends on `getInstalledAddons().first()`.

### Expected behavior

NuvioTV should replace the cached addon manifest when meaningful fetched manifest content changes, even if the top-level manifest `version` is unchanged. Home rows and addon reads should use the current catalog definitions, metadata, and ordered catalog descriptors.

### Actual behavior

The app can fetch a stale manifest but keep the old cached catalog definitions because cache replacement and home reload invalidation do not cover same-version manifest changes consistently. Before the fix, one-shot addon reads could return stale cached data before refresh, fresh cache-miss fetches could advance the global stale-refresh timer, and home rows could skip reloads when only config metadata, catalog descriptor fields, catalog order, or addon labels changed.

### Frequency

Always

### Did this work before?

Not sure

### Media details (optional)

This is addon manifest and home catalog cache behavior rather than a single media item.

### Logs (required for crash reports)

```shell
Not applicable; cache invalidation behavior.
```

### Anything else? (optional)

Replacement for #2454, which was closed by the unlabeled-issue cleanup. Fix PR: #2455.

### PR DESCRIPTION
**Pull request by dakotadawolfe**
## Summary

Updates addon manifest caching and related home reload behavior so NuvioTV refreshes cached addon definitions when meaningful manifest content changes, even if the top-level manifest `version` is unchanged.

The change also makes stale manifest refreshes run before stale cached data is emitted through `getInstalledAddons()`, keeps fresh cache-miss fetches from extending the global stale-refresh timer, and makes home catalog reloads notice addon labels plus ordered catalog descriptor changes.

## PR type

- [ ] Translation/localization only
- [x] Critical bug fix

## Why

Configurable Stremio addons can change catalog definitions for the same manifest URL without changing the top-level manifest `version`. For example, catalog `showInHome` flags, catalog lists, config versions, poster URLs, catalog paging, extras, catalog ordering, or addon labels may change after a user updates addon configuration.

Before this change, NuvioTV fetched stale manifests again but only replaced the cached manifest when `existing.version != addon.version`. If the version stayed the same, home rows could continue using old catalog definitions.

There were additional cache paths that could keep stale data visible after a refresh became necessary:

- one-shot callers using `getInstalledAddons().first()` could stop at the stale cached emission before the refresh ran
- a fresh cache-miss fetch could advance the global refresh timestamp even though existing cached manifests were not refreshed
- the home catalog load signature did not include addon `version` or `configVersion`, so existing home rows could be reused when only addon configuration metadata changed
- the home catalog load signature only included a subset of catalog fields and sorted catalog entries, so catalog descriptor changes or ordering changes could be missed
- addon label changes could be emitted by the repository while existing home rows kept the old `addonName`

## Issue or approval

Fixes #2466.

Replaces #2454, which was closed by the unlabeled-issue cleanup.

## Reproduction steps

1. Add a Stremio addon that exposes catalogs in its manifest.
2. Let NuvioTV load the addon and cache its manifest.
3. Update the same manifest URL so catalog definitions change while the manifest `version` stays the same. For example, change catalog `showInHome` flags, catalog poster URLs, catalog extras, page size, catalog order, or addon label.
4. Restart NuvioTV or wait for the manifest cache to become stale so the app fetches the manifest again.
5. Open the home screen.

Before this fix, the app could continue using the old cached catalog definitions, one-shot addon reads could stop at stale cached data, and home rows could skip reloads when only addon config metadata, catalog descriptor fields, catalog order, or addon labels changed. After this fix, meaningful manifest changes replace the cached manifest, stale refreshes feed back through `getInstalledAddons()`, the stale-refresh timer is only advanced by stale refreshes, and home rows reload when addon metadata, labels, or ordered catalog descriptor data changes.

## UI / behavior impact

- [x] No UI change
- [ ] No behavior change
- [ ] UI changed only to fix a documented glitch/bug
- [x] Behavior changed only to fix a documented bug/regression
- [ ] UI change has explicit maintainer approval
- [ ] Behavior change has explicit maintainer approval

## Policy check

- [x] I have read and understood `CONTRIBUTING.md`.
- [x] This PR fits the current PR policy: localization/translation only or a critical bug fix.
- [x] This PR does not add features, UI changes, refactors, or other non-critical changes.
- [x] This PR is small, focused, and limited to one issue.
- [x] This PR does not bundle unrelated refactors, cleanups, formatting, or drive-by changes.
- [x] This PR includes a linked issue, reproduction steps, and testing notes if it is a critical bug fix.
- [x] I listed the testing performed below.

> Feature additions, feature requests, UI changes, refactors, and other non-critical changes may be closed or deferred without review while NuvioTV is being prepared for a stable release.

## Scope boundaries

This does not change addon ordering controls, catalog rendering UI, UI layout, poster rendering, sync behavior, dependencies, architecture, schema, or the public addon contract.

The change is limited to addon manifest cache replacement, stale installed-addon emission order, stale-refresh timestamp handling, and home catalog reload invalidation when addon metadata, labels, or ordered catalog descriptor data changes.

Timestamp-only manifest changes are intentionally ignored so constantly changing `_timestamp` fields do not invalidate the cache by themselves.

## Testing

- Added `AddonManifestCachePolicyTest` coverage for:
  - same-version catalog visibility changes replacing the cached manifest
  - same-version `configVersion` changes replacing the cached manifest
  - timestamp-only changes not replacing the cached manifest
  - identical manifests not replacing cache state
  - stale cached manifests fetching before emitting to `first()` collectors
  - fresh cached manifests still emitting from cache without a fetch
  - stale refresh success advancing the refresh timestamp
  - fresh cache-miss fetches not advancing the refresh timestamp
- Added `HomeCatalogLoadSignatureTest` coverage for:
  - addon `configVersion` changes invalidating the home catalog load signature
  - addon `version` changes invalidating the home catalog load signature
  - addon `name` changes invalidating the home catalog load signature
  - addon `displayName` changes invalidating the home catalog load signature
  - catalog `extra` changes invalidating the home catalog load signature
  - catalog `pageSize` changes invalidating the home catalog load signature
  - catalog ordering changes invalidating the home catalog load signature
- Ran `git diff --check`.
- Ran `.\gradlew.bat :app:testFullDebugUnitTest --tests com.nuvio.tv.data.repository.AddonManifestCachePolicyTest --tests com.nuvio.tv.ui.screens.home.HomeCatalogLoadSignatureTest` locally with Android SDK 36 available.
- No emulator/device UI flow was run; this is non-UI repository/cache behavior covered by unit tests.

## Screenshots / Video

Not a UI change.

## Breaking changes

None.

## Linked issues

Fixes #2466.

### CODE REVIEW

File: app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt [hunk 1]
```diff
@@ -173,7 +156,7 @@ class AddonRepositoryImpl @Inject constructor(
                     val canonical = canonicalizeUrl(url)
                     (enabledByUrl[canonical] ?: true) && manifestCache[canonical] == null
                 }
-                if (hasCacheMiss) {
+                if (hasCacheMiss || isCacheStale()) {
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt]: Duh, yeah, I should have caught that path. Stale refresh now forces a manifest fetch instead of returning the cached entry first, and the refresh timestamp only moves forward after at least one fetch succeeds. Pushed in 70e314e.

[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt]: Silly human me

File: app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt [hunk 2]
```diff
@@ -373,3 +365,21 @@ class AddonRepositoryImpl @Inject constructor(
         }
     }
 }
+
+internal fun shouldReplaceCachedManifest(cached: Addon, fresh: Addon): Boolean {
+    if (cached.version != fresh.version) return true
+    if (cached.configVersion != fresh.configVersion) return true
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt:372]: Yep, this one checks out. configVersion can change without the catalog descriptors changing, so the home load signature now includes addon version/configVersion and has focused coverage for that case. Pushed in f280754.

File: app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt [hunk 3]
```diff
@@ -173,32 +156,41 @@ class AddonRepositoryImpl @Inject constructor(
                     val canonical = canonicalizeUrl(url)
                     (enabledByUrl[canonical] ?: true) && manifestCache[canonical] == null
                 }
-                if (hasCacheMiss) {
-                    val fresh = coroutineScope {
+                val cacheStale = isCacheStale()
+                if (hasCacheMiss || cacheStale) {
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt:161]: Yep, this checks out too. Stale cached manifests no longer emit before the refresh pass, so first() callers wait for the refreshed list or the cached fallback after the attempt. Covered in the repository cache policy test. Pushed in f280754.

**Comment by dakotadawolfe** [comment]
Hopefully this is more coverage. Idk what other hat trick edge case ai is going to find.

File: app/src/main/java/com/nuvio/tv/ui/screens/home/HomeViewModelCatalogUtils.kt [hunk 4]
```diff
@@ -11,10 +11,14 @@ internal fun HomeViewModel.catalogKey(addonId: String, type: String, catalogId:
...
 internal fun HomeViewModel.buildHomeCatalogLoadSignature(addons: List<Addon>): String {
+    return buildHomeCatalogLoadSignature(addons, disabledHomeCatalogKeys)
+}
+
+internal fun buildHomeCatalogLoadSignature(addons: List<Addon>, disabledHomeCatalogKeys: Set<String>): String {
     val addonCatalogSignature = addons
         .flatMap { addon ->
             addon.catalogs.map { catalog ->
-                "${addon.id}|${addon.baseUrl}|${catalog.apiType}|${catalog.id}|${catalog.name}|${catalog.showInHome}|${catalog.hasExplicitShowInHome}"
+                "${addon.id}|${addon.baseUrl}|${addon.version}|${addon.configVersion}|${catalog.apiType}|${catalog.id}|${catalog.name}|${catalog.showInHome}|${catalog.hasExplicitShowInHome}"
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/ui/screens/home/HomeViewModelCatalogUtils.kt]: Checked this and it was a real gap, though, extremely edge case. The home reload key now uses the full ordered catalog descriptor signature instead of only a subset, so catalog reorder, extra , pageSize, supported/required extras, etc. can invalidate existing home rows. Added focused tests for reorder, extra, and pageSize. Pushed in eda65b8.

File: app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt [hunk 5]
```diff
@@ -165,40 +148,50 @@ class AddonRepositoryImpl @Inject constructor(
...
+                                addon?.copy(enabled = enabled) to fetched
                             }
-                        }.awaitAll().filterNotNull()
+                        }.awaitAll()
                     }
+                    val fresh = refreshed.mapNotNull { it.first }
 
-                    if (fresh != cached) {
+                    if (refreshed.any { it.second }) {
+                        lastManifestRefreshTime = System.currentTimeMillis()
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/data/repository/AddonRepositoryImpl.kt]: Checked this path and fixed it. A fresh cache-miss fetch no longer advances the global manifest refresh time; that timestamp only moves after a stale refresh actually fetched at least one manifest. Added focused coverage for both cases. Pushed in 1813057.

File: app/src/main/java/com/nuvio/tv/ui/screens/home/HomeViewModelCatalogUtils.kt [hunk 6]
```diff
@@ -11,21 +12,55 @@ internal fun HomeViewModel.catalogKey(addonId: String, type: String, catalogId:
...
         .asSequence()
         .sorted()
         .joinToString(separator = ",")
     return "$addonCatalogSignature::$disabledSignature"
 }
 
+private fun Addon.homeCatalogSignature(): String {
+    val catalogSignature = catalogs
+        .joinToString(separator = ";") { catalog -> catalog.homeCatalogSignature() }
+    return "${id}|${baseUrl}|${version}|${configVersion}|${catalogSignature}"
```
[the remark this answers is not in the document]
Comment (dakotadawolfe) [reply app/src/main/java/com/nuvio/tv/ui/screens/home/HomeViewModelCatalogUtils.kt]: Checked this too. Home rows store `addon.displayName`, so the reload signature now includes addon `name` and `displayName` to catch label-only manifest/custom-name changes. Added tests for both. Pushed in 1813057.

**Comment by skoruppa** [comment]
Check if those changes are still needed after the latest update. And if so, resolve conflicts ;) thanks

**Comment by dakotadawolfe** [comment]
> Check if those changes are still needed after the latest update. And if so, resolve conflicts ;) thanks

Yep! I’ll take a look tomorrow and see :D
